src/main/objects/algorithms/SopranoGraphBuilder.js

```diff
@@ -0,0 +1,195 @@
+                    var nextNeighbour = currentNode.nextNeighbours[k].node;
+                    currentNode.nestedLayer.connectWith(nextNeighbour.nestedLayer, innerEvaluator, i===0, false);
+                }
+            }
+        }
+    }
+
+    var removeUselessNodesInNestedLayers = function(graph) {
+        //without last layer cause the first node is not present yet
+        for(var i=graph.layers.length-2; i>=0; i--) {
+            for (var j = 0; j < graph.layers[i].nodeList.length; j++) {
+                var currentNode = graph.layers[i].nodeList[j];
+                currentNode.nestedLayer.removeUselessNodes()
+            }
+        }
+    }
+
+    var removeUnreachableNodesInNestedLayers = function(graph) {
+        //without last layer cause the first node is not present yet
+        for(var i=1; i<graph.layers.length; i++) {
+            for (var j = 0; j < graph.layers[i].nodeList.length; j++) {
+                var currentNode = graph.layers[i].nodeList[j];
+                currentNode.nestedLayer.removeUnreachableNodes()
+            }
+        }
+    }
+
+    var removeNodesWithEmptyNestedLayers = function(graph) {
+        for(var i=graph.layers.length-1; i>=0; i--) {
+            for (var j = 0; j < graph.layers[i].nodeList.length; j++) {
+                var currentNode = graph.layers[i].nodeList[j];
+                if(currentNode.nestedLayer.isEmpty()){
+                    graph.layers[i].removeNode(currentNode);
+                    j--;
+                }
+            }
+        }
+    }
+
+    var removeUnreachableNodes = function (graph) {
+        for (var i = 0; i < graph.layers.length; i++) {
+            graph.layers[i].removeUnreachableNodes()
+        }
+    }
+
+    var removeUselessNodes = function (graph) {
+        for (var i = graph.layers.length - 1; i >= 0; i--) {
+            graph.layers[i].removeUselessNodes();
+        }
+    }
+
+    var setEdgeWeightsAndPropagate = function(graph, evaluator){
+        for(var i=0; i<graph.layers.length - 1; i++){
+            for(var j=0; j<graph.layers[i].nodeList.length; j++){
+                var currentNode = graph.layers[i].nodeList[j];
+
+                for(var k=0; k<currentNode.nextNeighbours.length; k++){
+                    var neighbour = currentNode.nextNeighbours[k];
+                    var connection = new RulesCheckerUtils.Connection(neighbour.node.content, currentNode.content)
+
+                    var w = evaluator.evaluateSoftRules(connection);
+                    neighbour.setWeight(w);
+
+                    //propagate
+                    var hf_left = currentNode.content.harmonicFunction;
+                    var hf_right = neighbour.node.content.harmonicFunction;
+
+                    for(var l=0; l<currentNode.nestedLayer.nodeList.length; l++){
+                        var nested_node = currentNode.nestedLayer.nodeList[l];
+                        for(var m=0; m<nested_node.nextNeighbours.length; m++){
```

Comment:
Te 5 forów mnie trochę martwi, ale podejrzewam, że inaczej nie dało rady. Ew możnaby te dwa najgłębsze do osobnej funkcji wynieść

Reply:
Wyniesione. Acz ech możnaby lepiej... ale to niewielki zysk będzie a ten refactor mnie przeraża